runtime/minor_gc.c

```diff
@@ -687,38 +692,59 @@ void caml_empty_minor_heap_promote(caml_domain_state* domain,
   /* leave the barrier */
   if( participating_count > 1 ) {
     CAML_EV_BEGIN(EV_MINOR_LEAVE_BARRIER);
-    {
-      SPIN_WAIT {
-        if (atomic_load_acquire(&domains_finished_minor_gc) ==
-            participating_count) {
-          break;
-        }
+    minor_gc_leave_barrier(domain, participating_count);
+    CAML_EV_END(EV_MINOR_LEAVE_BARRIER);
+  }
+}
 
-        caml_do_opportunistic_major_slice(domain, 0);
-      }
+static void minor_gc_leave_barrier
+  (caml_domain_state* domain, int participating_count)
+{
+  /* Spin while we have major work available */
+  SPIN_WAIT_BOUNDED {
+    if (caml_plat_barrier_is_released(&domains_finished_minor_gc)) {
+      return;
+    }
+
+    if (!caml_do_opportunistic_major_slice(domain, 0)) {
+      break;
     }
-    CAML_EV_END(EV_MINOR_LEAVE_BARRIER);
   }
+
+  /* Spin a bit longer, which is far less fruitful if we're waiting on more than
+     one thread */
+  unsigned spins = participating_count == 2
+    ? Max_spins_medium : Max_spins_short;
+  SPIN_WAIT_NTIMES(spins) {
+    if (caml_plat_barrier_is_released(&domains_finished_minor_gc)) {
+      return;
+    }
+  }
+
+  /* If there's nothing to do, block */
+  caml_plat_barrier_wait(&domains_finished_minor_gc);
 }
 
-void caml_do_opportunistic_major_slice
-  (caml_domain_state* domain_unused, void* unused)
+
+int caml_do_opportunistic_major_slice
```

Comment:
Unfortunately, the runtime code is not consistent in its use of `bool` versus `int`. This should be the topic for some discussion among maintainters, then another PR.

I do agree it would be better to add a comment.